GMS/FullGMS.cpp

```diff
@@ -0,0 +1,193 @@
+  gms.Identification->YearOfConstruction = 2022;
+  gms.Identification->SoftwareRevision = "v1.05";
+  gms.Identification->DeviceClass = "CoordinateMeasuringMachine";
+  gms.Identification->Location = "CTRL 5 5318/VIRTUAL 0 0/N 48.7685303 E 9.2653923";
+  gms.Identification->Model = {"", MachineName};
+
+  {
+    std::hash<std::string> hasher;
+    std::stringstream ss;
+    ss << "2022-" << hasher(MachineName);
+    gms.Identification->SerialNumber = ss.str();
+  }
+
+  InstantiateOptional(gms.Identification->SubDeviceClass, m_pServer, n);
+  InstantiateOptional(gms.Identification->Workspace, m_pServer, n);
+
+  gms.Identification->SubDeviceClass = "CoordinateMeasuringMachine";
+}
+
+void FullGMS::InstantiateMonitoring() {
+  InstantiateOptional(gms.Monitoring->MachineTool->PowerOnDuration, m_pServer, n);
+  InstantiateOptional(gms.Monitoring->MachineTool->FeedOverride, m_pServer, n);
+  gms.Monitoring->MachineTool->OperationMode = UA_MachineOperationMode::UA_MACHINEOPERATIONMODE_AUTOMATIC;
+}
+
+void FullGMS::InstantiateProduction() {
+  InstantiateOptional(gms.Production->ActiveProgram->State, m_pServer, n);
+  gms.Production->ActiveProgram->NumberInList = 0;
+  gms.Production->ActiveProgram->Name = "Basic Measurment Program";
+  gms.Production->ActiveProgram->State->CurrentState->Value = {"en", "Running"};
+  gms.Production->ActiveProgram->State->CurrentState->Number = 1;
+  gms.Production->ActiveProgram->State->CurrentState->Id =
+    UA_NODEID_NUMERIC(nsFromUri(m_pServer, constants::NsMachineToolUri), UA_MACHINETOOLID_PRODUCTIONSTATEMACHINETYPE_RUNNING);
+}
+
+void FullGMS::InstantiateEquipment() {
+  InstantiateOptional(gms.Equipment->Tools, m_pServer, n);
+
+  auto &tool1 = gms.Equipment->Tools->Tool.Add<machineTool::Tool_t>(m_pServer, n, {m_nsIndex, "Tool 1"});
+  tool1.ControlIdentifier1 = 12;
+  tool1.ControlIdentifierInterpretation = UA_ToolManagement::UA_TOOLMANAGEMENT_NUMBERBASED;
+  tool1.Name = "My Tool";
+  InstantiateOptional(tool1.Name, m_pServer, n);
+
+  auto &sensor1 = gms.Equipment->Tools->Tool.Add<GMS::GMSSensor_t>(m_pServer, n, {m_nsIndex, "Sensor 1"});
+
+  for (const auto &kv : CLASS_MAP) {
+    sensor1.Class->EnumStrings->push_back(kv.second);
+  }
+  sensor1.Class->Value = 2; /* TactileTouchTrigger */
+
+  sensor1.ControlIdentifier1 = 11;
+  sensor1.ControlIdentifier2 = 3;
+  sensor1.ControlIdentifierInterpretation = UA_ToolManagement::UA_TOOLMANAGEMENT_GROUPBASED;
+  sensor1.Locked->Value = true;
+  sensor1.Locked->ReasonForLocking = UA_ToolLocked::UA_TOOLLOCKED_BYOPERATOR;
+  sensor1.Name = "My Sensor";
+  InstantiateOptional(sensor1.ControlIdentifier2, m_pServer, n);
+  InstantiateOptional(sensor1.Name, m_pServer, n);
+
+  // Additional SensorType members
+  InstantiateOptional(sensor1.Alignment, m_pServer, n);
+  sensor1.Alignment = UA_ToolAlignmentState::UA_TOOLALIGNMENTSTATE_INDEXED;
+
+  InstantiateOptional(sensor1.Axes, m_pServer, n);
+  sensor1.Axes->push_back("X,Y,Z");
```

Comment:
This should be one entry per axe:
 sensor1.Axes->push_back("X");
 sensor1.Axes->push_back("Y");
 sensor1.Axes->push_back("Z");

Reply:
Ah, please take a look at the issue I posted [here](https://github.com/umati/GMS/issues/11)
Something is not right here or I'm implementing it wrong, but when I model this as a list I get exceptions about incompatible types.

Reply:
@Kantiran91 I closed the mentioned issue because I realized I was wrong and the nodeset xml was indeed correct. The Axes variable type has a ValueRank=1 (array) and ArrayDimensions=0 (unlimited size).
The actual problem must be somewhere else. 😕

Reply:
I have found the problem. At line 160 of `UmatiServerLib::Instantiation` all one-dimensional arrays were limited to length of 1. The value of the dimension(s) should be set to the max array length if known and to `0` if the array should be treated as an open-ended list.